File: 2023/day-8/fsharp/Program.fs
open System.IO
open System.Linq
open System

type Node(value, l, r) = 
    member this.Value:string = value
    member this.L:string = l
    member this.R:string = r

    member this.EndsWith(value:string) = 
        this.Value.EndsWith(value)

type NodeList(nodes_str) = 
    member this.Nodes = nodes_str |> List.map(fun(a:string) -> 
                        let parts = a.Replace("(", "").Replace(")", "").Split('=')
                        let value = parts.[0]
                        let lr = parts.[1].Split(',')
                        let l = lr.[0]
                        let r = lr.[1]
                        Node(value, l, r)
                    )
    member this.FindManyByValueEndsWith(value:string) = this.Nodes |> List.filter(fun x -> x.Value.EndsWith(value))
    member this.FindByValue(value:string) = this.Nodes.First(fun x -> x.Value = value)

let dirs, nodes_str = File.ReadAllLines("../../../input.txt") 
                        |> Array.toList 
                        |> function 
                            | instructions :: _ :: nodes_str  -> (instructions, nodes_str)
                            | _ -> failwith("Invalid input")

let node_list = NodeList(nodes_str)


let mutable current_nodes = node_list.FindManyByValueEndsWith("A")
let mutable current_step = 0
let mutable loop = true

while loop do
    if (current_nodes.All(fun x -> x.Value.EndsWith('Z')) = false) then
        loop <- false
    else
        current_nodes <- match dirs[current_step % dirs.Length] with
                            | 'L' -> current_nodes |> List.map(fun x -> node_list.FindByValue(x.L))
                            | 'R' -> current_nodes |> List.map(fun x -> node_list.FindByValue(x.R))
                            | _ -> failwith("Unkonwn dir")
        current_step <- current_step + 1

printfn($"{current_step}")
Console.ReadKey(true)

File: 2023/day-8/csharp/Program.cs
using System.Linq;

var lines = File.ReadAllLines("../../../input.txt");

var dirs = lines[0];
var nodes = lines.Skip(2).Select(Node.FromInput).ToList();
var node_dic = nodes.ToDictionary(x => x.Value);

var current_nodes = nodes.GetManyNodesByEndsWithValue("A");
long nr_of_loops = 0;
int current_step = 0;

while (!current_nodes.TrueForAll(x => x.Value.EndsWith("Z")))
{
    if (current_step >= dirs.Length){
        current_step = 0;
        nr_of_loops++;
    }
    current_nodes = dirs[current_step] switch {
        'R' => current_nodes.Select(x => node_dic[x.R]).ToList(),
        'L' => current_nodes.Select(x => node_dic[x.L]).ToList()
    };
    Console.WriteLine($"Steps: {GetTotalSteps()}");
    current_step++;
}


Console.WriteLine(GetTotalSteps());

long GetTotalSteps(){
    long total = nr_of_loops * dirs.Length + current_step;
    return total;
}

public static class Helper{
    public static Node GetNodeByValue(this List<Node> nodes, string value) => nodes.First(x => x.Value == value);
    public static List<Node> GetManyNodesByEndsWithValue(this List<Node> nodes, string value) => nodes.Where(x => x.Value.EndsWith(value)).ToList();
} 
public record Node(string Value, string L, string R){
    public static Node FromInput(string input){
        var values = input.Split('=');
        var value = values[0].Trim();
        var lr = values[1].Replace("(", "").Replace(")", "").Split(',');
        return new Node(value, lr[0].Trim(), lr[1].Trim());
    }
};